File: src/auth/auth.service.ts
import {
  Injectable,
  UnauthorizedException,
  ConflictException,
  BadRequestException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { PrismaService } from '../prisma/prisma.service';
import { RegisterDto } from './dto/register.dto';
import { LoginDto } from './dto/login.dto';
import * as bcrypt from 'bcrypt';
import { JwtPayload } from './strategies/jwt.strategy';

@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
  ) {}

  async register(registerDto: RegisterDto) {
    const { email, username, password, profileType, journeyType, systemSlug } =
      registerDto;

    const existingUser = await this.prisma.user.findFirst({
      where: {
        OR: [{ email }, { username }],
      },
    });

    if (existingUser) {
      throw new ConflictException('Email or username already exists');
    }

    const system = await this.prisma.system.findUnique({
      where: { slug: systemSlug },
    });

    if (!system) {
      throw new BadRequestException('Invalid system slug');
    }

    const hashedPassword = await bcrypt.hash(password, 10);

    const user = await this.prisma.user.create({
      data: {
        email,
        username,
        password: hashedPassword,
        profileType,
        journeyType,
        systemId: system.id,
      },
      include: {
        system: {
          select: {
            id: true,
            name: true,
            slug: true,
          },
        },
      },
    });

    const tokens = await this.generateTokens({
      sub: user.id,
      email: user.email,
      systemId: user.systemId,
    });

    await this.saveRefreshToken(user.id, tokens.refreshToken);

    return {
      user: {
        id: user.id,
        email: user.email,
        username: user.username,
        profileType: user.profileType,
        journeyType: user.journeyType,
        system: user.system,
      },
      ...tokens,
    };
  }

  async login(loginDto: LoginDto) {
    const { email, password } = loginDto;

    const user = await this.prisma.user.findUnique({
      where: { email },
      include: {
        system: {
          select: {
            id: true,
            name: true,
            slug: true,
          },
        },
      },
    });

    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const isPasswordValid = await bcrypt.compare(password, user.password);

    if (!isPasswordValid) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const tokens = await this.generateTokens({
      sub: user.id,
      email: user.email,
      systemId: user.systemId,
    });

    await this.saveRefreshToken(user.id, tokens.refreshToken);

    return {
      user: {
        id: user.id,
        email: user.email,
        username: user.username,
        profileType: user.profileType,
        journeyType: user.journeyType,
        system: user.system,
      },
      ...tokens,
    };
  }

  async refreshToken(refreshToken: string) {
    const storedToken = await this.prisma.refreshToken.findUnique({
      where: { token: refreshToken },
      include: {
        user: {
          include: {
            system: true,
          },
        },
      },
    });

    if (!storedToken) {
      throw new UnauthorizedException('Invalid refresh token');
    }

    if (storedToken.expiresAt < new Date()) {
      await this.prisma.refreshToken.delete({
        where: { id: storedToken.id },
      });
      throw new UnauthorizedException('Refresh token expired');
    }

    const tokens = await this.generateTokens({
      sub: storedToken.user.id,
      email: storedToken.user.email,
      systemId: storedToken.user.systemId,
    });

    await this.prisma.refreshToken.delete({
      where: { id: storedToken.id },
    });

    await this.saveRefreshToken(storedToken.user.id, tokens.refreshToken);

    return {
      user: {
        id: storedToken.user.id,
        email: storedToken.user.email,
        username: storedToken.user.username,
        profileType: storedToken.user.profileType,
        journeyType: storedToken.user.journeyType,
        system: storedToken.user.system,
      },
      ...tokens,
    };
  }

  async logout(userId: string) {
    await this.prisma.refreshToken.deleteMany({
      where: { userId },
    });

    return { message: 'Logged out successfully' };
  }

  private async generateTokens(payload: JwtPayload) {
    const accessToken = this.jwtService.sign(payload, {
      secret: this.configService.get<string>('jwt.secret'),
      expiresIn: this.configService.get<string>('jwt.expiresIn'),
    });

    const refreshToken = this.jwtService.sign(payload, {
      secret: this.configService.get<string>('refreshToken.secret'),
      expiresIn: this.configService.get<string>('refreshToken.expiresIn'),
    });

    return { accessToken, refreshToken };
  }

  private async saveRefreshToken(userId: string, token: string) {
    const expiresIn =
      this.configService.get<string>('refreshToken.expiresIn') || '7d';
    const expirationDate = new Date();

    const days = parseInt(expiresIn.replace('d', ''));
    expirationDate.setDate(expirationDate.getDate() + days);

    await this.prisma.refreshToken.create({
      data: {
        userId,
        token,
        expiresAt: expirationDate,
      },
    });
  }
}

File: app/services/auth_service.py
from datetime import datetime, timedelta
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from fastapi import HTTPException, status

from app.models.user import User, RefreshToken
from app.models.system import System
from app.schemas.auth import RegisterRequest, LoginRequest, AuthResponse, UserResponse, SystemResponse
from app.core.security import (
    verify_password,
    get_password_hash,
    create_access_token,
    create_refresh_token,
    decode_refresh_token,
    parse_expires_in
)
from app.core.config import settings


class AuthService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def register(self, data: RegisterRequest) -> AuthResponse:
        result = await self.db.execute(
            select(User).where((User.email == data.email) | (User.username == data.username))
        )
        existing_user = result.scalar_one_or_none()

        if existing_user:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Email or username already exists"
            )

        result = await self.db.execute(
            select(System).where(System.slug == data.systemSlug)
        )
        system = result.scalar_one_or_none()

        if not system:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid system slug"
            )

        hashed_password = get_password_hash(data.password)

        user = User(
            email=data.email,
            username=data.username,
            password=hashed_password,
            profile_type=data.profileType,
            journey_type=data.journeyType,
            system_id=system.id,
        )

        self.db.add(user)
        await self.db.commit()
        await self.db.refresh(user)

        result = await self.db.execute(
            select(System).where(System.id == user.system_id)
        )
        user_system = result.scalar_one()

        tokens = await self._generate_tokens(user)
        await self._save_refresh_token(user.id, tokens["refreshToken"])

        return AuthResponse(
            user=UserResponse(
                id=user.id,
                email=user.email,
                username=user.username,
                profileType=user.profile_type,
                journeyType=user.journey_type,
                system=SystemResponse(
                    id=user_system.id,
                    name=user_system.name,
                    slug=user_system.slug,
                )
            ),
            accessToken=tokens["accessToken"],
            refreshToken=tokens["refreshToken"],
        )

    async def login(self, data: LoginRequest) -> AuthResponse:
        result = await self.db.execute(
            select(User).where(User.email == data.email)
        )
        user = result.scalar_one_or_none()

        if not user or not verify_password(data.password, user.password):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid credentials"
            )

        result = await self.db.execute(
            select(System).where(System.id == user.system_id)
        )
        user_system = result.scalar_one()

        tokens = await self._generate_tokens(user)
        await self._save_refresh_token(user.id, tokens["refreshToken"])

        return AuthResponse(
            user=UserResponse(
                id=user.id,
                email=user.email,
                username=user.username,
                profileType=user.profile_type,
                journeyType=user.journey_type,
                system=SystemResponse(
                    id=user_system.id,
                    name=user_system.name,
                    slug=user_system.slug,
                )
            ),
            accessToken=tokens["accessToken"],
            refreshToken=tokens["refreshToken"],
        )

    async def refresh(self, refresh_token: str) -> AuthResponse:
        result = await self.db.execute(
            select(RefreshToken).where(RefreshToken.token == refresh_token)
        )
        stored_token = result.scalar_one_or_none()

        if not stored_token:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid refresh token"
            )

        if stored_token.expires_at < datetime.utcnow():
            await self.db.delete(stored_token)
            await self.db.commit()
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Refresh token expired"
            )

        result = await self.db.execute(
            select(User).where(User.id == stored_token.user_id)
        )
        user = result.scalar_one()

        result = await self.db.execute(
            select(System).where(System.id == user.system_id)
        )
        user_system = result.scalar_one()

        tokens = await self._generate_tokens(user)

        await self.db.delete(stored_token)
        await self._save_refresh_token(user.id, tokens["refreshToken"])
        await self.db.commit()

        return AuthResponse(
            user=UserResponse(
                id=user.id,
                email=user.email,
                username=user.username,
                profileType=user.profile_type,
                journeyType=user.journey_type,
                system=SystemResponse(
                    id=user_system.id,
                    name=user_system.name,
                    slug=user_system.slug,
                )
            ),
            accessToken=tokens["accessToken"],
            refreshToken=tokens["refreshToken"],
        )

    async def logout(self, user_id: str) -> dict:
        result = await self.db.execute(
            select(RefreshToken).where(RefreshToken.user_id == user_id)
        )
        tokens = result.scalars().all()

        for token in tokens:
            await self.db.delete(token)

        await self.db.commit()
        return {"message": "Logged out successfully"}

    async def _generate_tokens(self, user: User) -> dict:
        access_expires = parse_expires_in(settings.JWT_EXPIRES_IN)
        refresh_expires = parse_expires_in(settings.REFRESH_TOKEN_EXPIRES_IN)

        access_token = create_access_token(
            data={"sub": user.id, "email": user.email, "systemId": user.system_id},
            expires_delta=access_expires
        )

        refresh_token = create_refresh_token(
            data={"sub": user.id, "email": user.email, "systemId": user.system_id},
            expires_delta=refresh_expires
        )

        return {
            "accessToken": access_token,
            "refreshToken": refresh_token,
        }

    async def _save_refresh_token(self, user_id: str, token: str):
        expires_in = parse_expires_in(settings.REFRESH_TOKEN_EXPIRES_IN)
        expiration_date = datetime.utcnow() + expires_in

        refresh_token = RefreshToken(
            user_id=user_id,
            token=token,
            expires_at=expiration_date,
        )

        self.db.add(refresh_token)
        await self.db.commit()
